Follow the behaviour of the C compiler for the `.size` and `.type` directives and the `.note.GNU-stack` section

shym opened the pull request:
This PR makes the decision of adding a `.note.GNU-stack` section in the assembler code uniform by detecting once and for all in `configure` whether it is supported and always check the result of that detection to add the section.

Before that PR, the `.note.GNU-stack` section was:
- on x86_64:
  - generated by `emit.mlp` only on Linux and FreeBSD
  - present in `amd64.S` only on Linux and FreeBSD
- on ARM:
  - generated by `emit.mlp` only on Linux
  - present in `arm64.S` except on macOS
- on RISC-V:
  - never generated by `emit.mlp`
  - always present in `riscv.S`
- on POWER:
  - always generated by `emit.mlp`
  - always present in `power.S`
- on S/390:
  - always generated by `emit.mlp`
  - always present in `s390x.S`

Besides making the decision uniform, this PR proposes to change the logic of the decision so that the `.note.GNU-stack` section is always output if the toolchain supports it. After evaluating a couple of options for that, I propose to simply test whether the object format is ELF as:

- the GNU linker requires that note for ELF binaries: without the note, it marks the stack as executable and displays a warning that this is deprecated (1),
- the LLVM linker supports that note for ELF binaries: it ignores it, it always marks the stack as non-executable (this has been tested on FreeBSD/x86_64, clang+LLD 18.1.5; note that in that case, `clang -S` generates the section even if LLD will ignore it),
- the note is not supported when generating other formats, even with the GNU linker.

This has been tested on the combinations GCC+GNU ld on Linux x86_64, clang+LLD on FreeBSD x86_64, GCC+GNU ld on FreeBSD x86_64 and what we test in CI.

The need for this PR arose from the MirageOS use case, where checking for the OS type was not yielding the necessary result.

(1) For instance, GNU ld v2.43.50 shows:
```
/usr/bin/ld: warning: .... missing .note.GNU-stack section implies executable stack
/usr/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker
```

shym committed:
Detect the need for the GNU note for non-executable stack

Detect during `configure` whether the C compiler generates a
`.note.GNU-stack` section to mark it explicitly as non-executable
Define a macro in `m.h` and a configure variable so that OCaml can do as
cc does

Detail: without the explicit `.note.GNU-stack` section, recent versions
of GNU ld show, at least on some targets:
/usr/bin/ld: warning: .... missing .note.GNU-stack section implies executable stack
/usr/bin/ld: NOTE: This behaviour is deprecated and will be removed in a future version of the linker

shym committed:
Add a `Config` variable for the need of the GNU non-executable stack note

Add a `Config.with_nonexecstack_note` variable to indicate whether the
`.note.GNU-stack` section should be present to mark the stack explicitly
as non-executable

shym committed:
Output the `.note.GNU-stack` section based on `configure`

Uniformly rely on the detection during `configure` of whether the
`.note.GNU-stack` section is generated by the C compiler to decide
whether to output it
Define a macro in `asm.h` and a function in `Emitaux` to output the
section when it should be

Before that patch, the `.note.GNU-stack` section was:
- on x86_64:
  - generated by `emit.mlp` only on Linux and FreeBSD
  - present in `amd64.S` only on Linux and FreeBSD
- on ARM:
  - generated by `emit.mlp` only on Linux
  - present in `arm64.S` except on macOS
- on RISC-V:
  - never generated by `emit.mlp`
  - always present in `riscv.S`
- on POWER:
  - always generated by `emit.mlp`
  - always present in `power.S`
- on S/390:
  - always generated by `emit.mlp`
  - always present in `s390x.S`

MisterDA commented:
I'm tempted to say that you don't need to introduce the `SUPPORT_NOTE_GNU_STACK_NOEXEC` macro, you can directly test for `#ifdef __ELF__` in C and assembly files, as you're doing in the configure script. Maybe the assembler doesn't define this macro?

shym commented:
> I'm tempted to say that you don't need to introduce the `SUPPORT_NOTE_GNU_STACK_NOEXEC` macro, you can directly test for `#ifdef __ELF__` in C and assembly files, as you're doing in the configure script. Maybe the assembler doesn't define this macro?

I didn’t test whether `__ELF__` is defined in assembler but anyway we would lose one important goal of this PR, which is to make the decision in a single point of the code.

dustanddreams commented on configure.ac:
```diff
@@ -676,6 +677,20 @@ popdef([host_cpu])dnl
+# supports it in the sense that it ignores it, always marking the stack as
+# non-executable)
+AC_PREPROC_IFELSE([AC_LANG_SOURCE([[
+  #ifndef __ELF__
```
I find this test horribly misleading.

What you are testing here, is whether the system you're building on uses ELF binaries.
While this is relevant to the "GNU-stack" note since notes are ELF-specific, not all ELF platforms or toolchains actually need these notes.

First, systems using clang's linker instead of GNU Binutils `ld`, do not need these notes, because control of the stack's executability permission is performed using other options. Quoting from clang's `lld/ELF/InputFiles.cpp`:
```
    // The GNU linker uses .note.GNU-stack section as a marker indicating
    // that the code in the object file does not expect that the stack is
    // executable (in terms of NX bit). If all input files have the marker,
    // the GNU linker adds a PT_GNU_STACK segment to tells the loader to
    // make the stack non-executable. Most object files have this section as
    // of 2017.
    //
    // But making the stack non-executable is a norm today for security
    // reasons. Failure to do so may result in a serious security issue.
    // Therefore, we make LLD always add PT_GNU_STACK unless it is
    // explicitly told to do otherwise (by -z execstack). Because the stack
    // executable-ness is controlled solely by command line options,
    // .note.GNU-stack sections are simply ignored.
```

Second, some systems such as NetBSD and OpenBSD do not allow for the stack to be executable, regardless of whatever linker trickery you'll try to do. On these systems, there is no need to create such notes either.

The conditions for these notes to be useful are, to the best of my knowledge:
- object binary format is ELF.
- system is either Linux or FreeBSD (or, system is neither DragonflyBSD nor NetBSD nor OpenBSD).
- linker is GNU `ld`.

shym replied:
Thank you for taking the time to look into this.

The intention of this test is not to detect when the `.note.GNU-stack` sections are actually useful but rather to make sure that:
1. the sections are always present when they are needed,
2. the sections are always absent when they would break things,
3. the test is as simple as possible in cases where it doesn’t matter, namely when it will work either way. I just checked what `clang -S` does on OpenBSD 7.6 x86_64: the section is present in that case also. On the NetBSD system I could check, `cc` is GCC 10 where the note is not generated. I don’t have access to a DragonflyBSD.

So I wonder:
- is there any 1. or 2. case that just checking for `__ELF__` doesn’t catch?
- should we aim at creating the section really only when needed?
- should the comment and/or variable names be chosen so that the purpose and limits of the test are more explicit?

(For what it’s worth, a first iteration of this PR tried to test whether we are using the GNU linker; I changed my mind when I saw in the cases I tested that `autoconf` identified LLD as GNU ld (because `lld -v` says it is compatible with GNU linkers) and I was not so sure it was a relevant condition, as clang also generated the section)

dustanddreams replied:
I would myself prefer these sections to be only output when they actually matter but I can understand if you prefer to aim for "only output when they don't risk breaking anything", which is easier to figure out.

However in any case, if you keep the test unchanged, then by all means give it a name which matches what it does, something like `elf_toolchain`, not `linker_support_note_gnu_stack_noexec`.

shym committed:
Add a Changes entry

shym commented:
I just pushed a revised version thanks to @dustanddreams’ review.

Changes:
- `configure` will now run the C compiler and see if it generates a `.note.GNU-stack` section in its assembler code; when in Rome, do as the C compilers do?
- the variable name is `generate_noexecstack_note`, I hope it makes more sense,
- the test is properly logged and its result is cached.

(Tested on {Open,Net,Free}BSD x86_64 in addition to the CI runs)

dustanddreams reviewed (approved):
Looking good and much less error-prone now.

gasche commented:
If you add a new value to `config.mli`, wouldn't it be nice to also update the `-config` and `-config-var` stuff, that is, the definition of `configuration_variables` in config.ml?

shym commented:
> If you add a new value to `config.mli`, wouldn't it be nice to also update the `-config` and `-config-var` stuff, that is, the definition of `configuration_variables` in config.ml?

As not all configuration keys are exposed that way, I supposed they are exposed only when a use case for them has been identified (and I have no use case in mind, at the moment at least). Did I suppose wrongly?

gasche commented:
I don't see a cost for exposing new variables, and I see a problem when people realize there is a use-case, but they cannot actually get the information because the OCaml versions they support do not expose the variable.

shym commented:
> I don't see a cost for exposing new variables, and I see a problem when people realize there is a use-case, but they cannot actually get the information because the OCaml versions they support do not expose the variable.

I agree. So I’ve added it after the other asm configuration variables and renamed it to `nonexecstack_note` to make it more alike other variable names (ie not a `verb_xyz` form).

MisterDA commented on aclocal.m4:
```diff
@@ -527,6 +527,30 @@ AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
+
+    ocaml_cv_prog_cc_nonexecstack_note=no
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE],
+      [AS_IF([$FGREP .note.GNU-stack conftest.$ac_objext >/dev/null],
```
```suggestion
      [AS_IF([$FGREP -q .note.GNU-stack conftest.$ac_objext],
```

shym replied:
I chose to follow how other uses of `$?GREP` are silenced in the rest of `configure`, to avoid any unexpected surprise.

shym committed:
Detect support for `.size` and `.type` assembler directives

Detect during `configure` whether the C compiler generates `.size` and
`.type` assembler directives
Set the result to true if and only if both directives are generated for
functions and the `.type` directive is generated for objects (directives
equivalent to `.size` could be used for a simple `int`); display a
warning if some of the three combinations appear but not all three
Define a macro in `m.h` and a configure variable so that OCaml can do as
cc does

shym committed:
Add a `Config` variable for the `.size` and `.type` assembler directives

Add a `Config.asm_size_type_directives` variable to indicate whether the
`.size` and `.type` assembler directives can be used in assembler code

shym committed:
Output the `.size` and `.type` directives based on `configure` result

Uniformly rely on the detection during `configure` of whether the
`.size` and `.type` directives are generated by the C compiler to decide
whether to output them
Define two macros in `asm.h` and two functions in `Emitaux` to output
those directives to factorize them; to do so, `Emitaux.emit_symbol`
knows the special macOS case

Before that patch, the `.size` and `.type` directives were:
- on x86_64:
  - generated by `emit.mlp` only on Linux and GNU OSes; `.size` was
    generated for frametables only for a definite list of OSes, `.type`
    was never generated for frametables
  - present in `amd64.S` only on Linux and GNU OSes; the `.type`
    directives was present for functions on all but macOS, Cygwin,
    MinGW, MSVC
- on ARM:
  - generated by `emit.mlp` except on macOS
  - present in `arm64.S` except on macOS
- on RISC-V:
  - always generated by `emit.mlp`; `.size` was not generated for
    frametables
  - always present in `riscv.S`
- on POWER:
  - always generated by `emit.mlp`; `.size` was not generated for
    frametables
  - always present in `power.S`
- on S/390:
  - always generated by `emit.mlp` for `.type`, never generated for
    `.size`
  - always present in `s390x.S` for `.type`, never present for `.size`

shym commented:
I learned last week that I was overly conservative about when to bootstrap: as the changes to the configuration that PR brings are not used in the bootstrap compiler, the bootstrap compiler doesn’t really need to be updated. So I’ve removed the bootstrap step.

I’ve also run into a problem due to the missing `.size` directive, again in a MirageOS context. So I’ve proceeded exactly the same way to fix it, namely by checking what the C compiler does. More precisely:
- `configure` contains a new test that looks whether `.size` and `.type` directives are generated,
- `Config` contains a new variable, `asm_size_type_directives`, for that result,
- I checked all the places where those directives were, or were not, generated, to rely uniformly on the test, I hope I didn’t miss one. (That commit ended up larger than I expected),
- I added the 3 missing Linux `x86_64`-to-* cross compilers in the CI test, to get some (at least minimal) testing of all 5 backends in the GHA.

As the changes for the `.size` and `.type` directives are so similar to the ones for non-executable stack note, I’ve added them to that PR and updated the title, only in `Changes` for now. They could equally be spawned as a separate PR.

I’ve also tested the branch on `x86_64` {Free,Open,Net}BSDs. Interestingly, NetBSD is the only system where I’ve observed the two tests return different results (no `.note.GNU-stack` section, but `.size` and `.type` directives).

gasche commented on asmcomp/arm64/emit.mlp:
```diff
@@ -54,13 +54,13 @@ let emit_symbol s =
-  if not macosx then begin
+  if Config.asm_size_type_directives then begin
     `	.size	{emit_symbol sym}, .-{emit_symbol sym}\n`
   end
```
Could/should these use the Emitaux helpers, to be more consistent with other backends?

shym replied:
I considered it and didn’t do it (yet) as `arm64` overrides `emit_symbol`. I think the nicest way to handle this would be to modify the common `emit_symbol` to add the `_` on macOS.
For the `.type` directive, the `arm64` implementation is even more convoluted, as it emits `.type` for some labels (I wondered when that is useful). The idea I had (without implementing it yet) would be to use a wrapper for `.type` on labels and use the `Emitaux` one directly for other use cases. That would rely again on adding `_` to the general `emit_symbol`.

shym replied:
Done in the last push.

gasche commented on asmcomp/arm64/emit.mlp:
```diff
@@ -1273,9 +1273,7 @@ let end_assembly () =
-  | _ -> ()
+  if Config.nonexecstack_note then begin
+    (* Mark stack as non-executable *)
+    `	.section .note.GNU-stack,\"\",%progbits\n`
```
Maybe we could have a `emit_nonexexstack_note` helper in Emitaux to hide the Config check here?

shym replied:
I just waited for an explicit suggestion to do so :-)

shym replied:
Done.

gasche commented on runtime/caml/asm.h:
```diff
@@ -45,6 +45,17 @@
+#define SIZE_DIRECTIVE(name)
+#define TYPE_DIRECTIVE(name,ty)
+#endif
+
```
Similarly here we could have a `ASM_NONEXECSTACK_NOTE` that is defined conditionally, to avoid the conditional at use-sites.

shym replied:
I just waited for an explicit suggestion to do so also :-)

shym replied:
Done.

gasche reviewed (commented):
Currently the size and type emissions are factorized in helper functions (most of the time), which is nice as it hides the conditional configuration check from the main files. I think you could do this consistently, and use the same approach for the stack note.

I can't review the configure stuff. I am inclined to just trust you on it (+ the CI apparently working just fine) if we don't have a review for it.

MisterDA commented on aclocal.m4:
```diff
@@ -527,6 +527,78 @@ AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
+    ocaml_cv_prog_cc_asm_size_type_directives=no
+    AC_COMPILE_IFELSE([AC_LANG_SOURCE([[
+int feat_detect_obj;
+int feat_detect_func() {
```
```suggestion
int feat_detect_func(void) {
```

MisterDA commented on aclocal.m4:
```diff
@@ -527,6 +527,78 @@ AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
+        [asm_size_func_directive=yes])
+      AS_CASE([m4_join([,],[$asm_type_obj_directive],[$asm_type_func_directive],
+          [$asm_size_func_directive])],
+        [yes,yes,yes],
+          [ocaml_cv_prog_cc_asm_size_type_directives=yes],
+        [no,no,no],
```
[Sí-Sí, No-No · Machito & His Orchestra · Graciela](https://www.youtube.com/watch?v=h3cg1luIFMo) ;)

shym committed:
Include `asm.h` at the beginning of assembler sources

gasche reviewed (approved):
This looks good to me, but I don't understand the .github/workflows/build-cross part of the changes. Are those new CI tests that will run on each PR? Do we want them? Or was it something you did for easier testing? Why are they kept in the PR?
(There is really a lot of duplication among cross-compilation CI scripts, I wish it was possible to factorize them.)

shym commented:
> This looks good to me, but I don't understand the .github/workflows/build-cross part of the changes. Are those new CI tests that will run on each PR?

Those tests run only on PRs with the `run-crosscompiler-tests` label. I’ve set it in my [self-PR](https://github.com/shym/ocaml/pull/23) to use them.

> Do we want them? 

I’m not sure, mostly because I don’t know what `precheck` does; I assume it can run the real test suite on all platforms, so they are probably not that useful. I just dropped them.

> Or was it something you did for easier testing?

Yes.

> (There is really a lot of duplication among cross-compilation CI scripts, I wish it was possible to factorize them.)

Indeed. A `tools/ci` helper script could help, but that’s not perfect either.

gasche commented:
Merged, thanks!
